src/vivarium_nih_moud/data/builder.py

```diff
@@ -127,3 +128,77 @@ def write_data_by_draw(artifact: Artifact, key: str, data: pd.DataFrame):
         data = data.reset_index(drop=True)
         for c in data.columns:
             store.put(f"{key.path}/{c}", data[c])
+
+
+def generate_consistent_moud_rates(art: Artifact, location: str, years: Optional[str]):
+    """Generates consistent rates for MOUD data.
+
+    Parameters
+    ----------
+    art
+        The artifact to read from and write to.
+    location
+        The location associated with the data to load and the artifact to
+        write to.
+    years
+        The years to load data for.
+
+    """
+    # TODO: check if the consistent rates are already in the artifact, and if so, skip rest of this function
+
+    ages = np.arange(0, 96, 5)
+    years = np.array([2020, 2025])
+    sexes = ["Male", "Female"]
+    key = {
+        "i": "cause.opioid_use_disorders.incidence_rate",
+        "p": "cause.opioid_use_disorders.prevalence",
+        "f": "cause.opioid_use_disorders.excess_mortality_rate",
+        "m_all": "cause.all_causes.cause_specific_mortality_rate",
+        "csmr_with": "cause.opioid_use_disorders.cause_specific_mortality_rate",
+        "pop": "population.structure",
+    }
+
+    def oud_data(sex):
+        df_data = pd.concat(
+            [
+                dismod_at.transform_to_data("p", art.load(key["p"]), sex, ages, [2021]),
+                dismod_at.transform_to_data("i", art.load(key["i"]), sex, ages, [2021]),
+                dismod_at.transform_to_data("f", art.load(key["f"]), sex, ages, [2021]),
+                dismod_at.transform_to_data(
+                    "m",
+                    art.load(key["m_all"]) - art.load(key["csmr_with"]),
+                    sex,
+                    ages,
+                    [2021],
+                ),
+            ]
+        )
+        return df_data
+
+    def get_rates(model_dict, rate_type, year):
+        df_out = []
+        for model in model_dict.values():
+            df_out.append(model.get_rate(rate_type, year))
+        df_out = pd.concat(df_out)
+        return df_out
+
+    # fit model separately for Male and Female
+    m = {}
+    for sex in sexes:
+        m[sex] = dismod_at.ConsistentModel(sex, ages, years)
+        m[sex].fit(oud_data(sex))
+
+    # store consistent rates in artifact
+    for rate_type in "ipfr":
+        # generate data for k
+        df_out = get_rates(m, rate_type, 2020)
+        # store generated data in artifact
+        if rate_type != "r":
+            rate_name = key[rate_type]
+        else:
+            rate_name = "cause.opioid_use_disorders.remission_rate"
+        rate_name += "_consistent"
+        if rate_name in art.keys:
+            art.replace(rate_name, df_out)
+        else:
+            art.write(rate_name, df_out)
```

Comment:
Is there a reason you put this in the builder.py and not loader.py?

Reply:
> Is there a reason you put this in the builder.py and not loader.py?

based on [this slack discussion](https://ihme.slack.com/archives/C221EE33J/p1728499702362699), it seemed most convenient (since [it is available in make_artifacts](https://github.com/ihmeuw/vivarium_nih_moud/blob/ac0b900c659fa2b6039e3c88ff956ce08ee5d8ca/src/vivarium_nih_moud/tools/make_artifacts.py#L238)). How should I think about the difference between builder and loader?